src/mono/mono/metadata/marshal-lightweight.c

```diff
@@ -2550,9 +2550,18 @@ emit_managed_wrapper_ilgen (MonoMethodBuilder *mb, MonoMethodSignature *invoke_s
 	 * return ret;
 	 */
 
+	/* delete_old = FALSE */
 	mono_mb_emit_icon (mb, 0);
 	mono_mb_emit_stloc (mb, 2);
 
+	/*
+	* Transformed into a direct icall when runtime init callback is enabled for a native-to-managed wrapper.
+	* This icall is special cased in the JIT so it can be called in native-to-managed wrapper before
+	* runtime has been initialized. On return, runtime must be fully initialized.
+	*/
+	if (runtime_init_callback)
+		mono_mb_emit_icall (mb, mono_dummy_runtime_init_callback);
```

Comment:
I know it's a huge pain to make a new `WrapperSubtype` and use the `MonoMethodBuilder` to construct it, but man, this really feels like a completely separate concern from the normal n2m wrapper.

What would be better (if it wasn't kind of a pain to do) is to wrap the n2m wrapper in a new wrapper whose only job is to invoke the runtime init callback and then invoke the normal wrapper.

And then let the inliner combine them into one function.

Reply:
I think my preferences are:
1. special "runtime init wrapper" wrapper around the n2m wrapper.  (As you said - too much work for little gain.  but architecturally it's the cleanest idea)
2. what you currently have - added code in the n2m wrapper. It's fine, it's still pretty clean.
3. messing with thread_attach.  I really don't like this option - it's really abusing an unrelated mechanism.

Reply:
OK, then I keep what we have in this PR.